Introduction of Edge Dictionaries

trentrosenbaum opened the pull request:
This PR represents an implementation of the Fastly Edge Dictionaries resource.  

The resource has two parts to it.  
- Association with the service itself.  One or more Edge Dictionaries can be created as nested blocks and are linked to a service version.  The configuration is activated in the same way as other service fields.  
- Versionless items that can be entered in to a dictionary.  These items are represented as a separate resource that references the dictionary from the service.  Within this resource a map representing the item values is supplied.

The resource that represents the items is called ```fastly_service_dictionary_items_v1``` and there will be one for every dictionary that requires items to be associated with it.  

The resource will only keep track of the items that it has created in the remote state.  The aim is to allow additional items to be added via external means, (e.g. api or console) and to not be removed by terraform.  If there is a need to update terraform created items, then the resource can be configured with the lifecycle ignore_change configuration and this will instruct updates to these items to be ignored.
```
lifecycle {
  ignore_changes = [items,]
}
```
The main design decision made is to represent the items as a single map within a resource instead of one resource per item.  The main positive is a reduction in repeated fields and that all related items are managed as a single grouping within Terraform.

Commit:
Initial implementation of dictionary and dictionary items

Commit:
Updates the service documentation to decribe the new dictionary block.

Commit:
Updates the website with initial page covering the new service dictionary items resource.

Commit:
Udpates the acceptance test with a refined expression to select the dictionary_id and show the use of variables to reduce repeating values.

Commit:
Adds two examples of how to use the service/dictionary and the dictionary items

Commit:
Updates the resource to update the items, (removal, update, create)

Commit:
Refactor the read function to filter out external dict items and implement acceptance tests.

Commit:
Removed the use of spliting the resource id to determine the service and dictionary ids.

Commit:
Adds acceptance test to confirm an update to dictionary name.

Commit:
Removes the write_only field from service dictionary nested block.

Commit:
Removed the extra HCL charatcers that are not needed for HCL2

Commit:
Updated acceptance tests and website docs with configuration showing 0.12.x features.

trentrosenbaum commented:
<details><summary>Acceptance Test Results</summary>
<p>
The following shows the result of the acceptance test execution against this branch/PR

```bash
$ make testacc
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v  -timeout 120m -ldflags="-X=github.com/terraform-providers/terraform-provider-fastly/version.ProviderVersion=acc"
?       github.com/terraform-providers/terraform-provider-fastly        [no test files]
=== RUN   TestUserAgentContainsProviderVersion
--- PASS: TestUserAgentContainsProviderVersion (0.00s)
=== RUN   TestAccFastlyIPRanges
--- PASS: TestAccFastlyIPRanges (2.39s)
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestResourceFastlyFlattenDictionaryItems
--- PASS: TestResourceFastlyFlattenDictionaryItems (0.00s)
=== RUN   TestAccFastlyServiceDictionaryItemV1_create
--- PASS: TestAccFastlyServiceDictionaryItemV1_create (45.27s)
=== RUN   TestAccFastlyServiceDictionaryItemV1_create_dynamic
--- PASS: TestAccFastlyServiceDictionaryItemV1_create_dynamic (44.62s)
=== RUN   TestAccFastlyServiceDictionaryItemV1_update
--- PASS: TestAccFastlyServiceDictionaryItemV1_update (103.94s)
=== RUN   TestAccFastlyServiceDictionaryItemV1_external_item_is_ignored
--- PASS: TestAccFastlyServiceDictionaryItemV1_external_item_is_ignored (69.01s)
=== RUN   TestAccFastlyServiceDictionaryItemV1_external_item_remains
--- PASS: TestAccFastlyServiceDictionaryItemV1_external_item_remains (86.22s)
=== RUN   TestResourceFastlyFlattenBigQuery
--- PASS: TestResourceFastlyFlattenBigQuery (0.17s)
=== RUN   TestAccFastlyServiceV1_bigquerylogging
--- PASS: TestAccFastlyServiceV1_bigquerylogging (36.09s)
=== RUN   TestAccFastlyServiceV1_bigquerylogging_env
--- PASS: TestAccFastlyServiceV1_bigquerylogging_env (34.36s)
=== RUN   TestResourceFastlyFlattenBlobStorage
--- PASS: TestResourceFastlyFlattenBlobStorage (0.00s)
=== RUN   TestAccFastlyServiceV1_blobstoragelogging_basic
--- PASS: TestAccFastlyServiceV1_blobstoragelogging_basic (78.05s)
=== RUN   TestAccFastlyServiceV1_blobstoragelogging_default
--- PASS: TestAccFastlyServiceV1_blobstoragelogging_default (35.09s)
=== RUN   TestAccFastlyServiceV1_blobstoragelogging_env
--- PASS: TestAccFastlyServiceV1_blobstoragelogging_env (36.20s)
=== RUN   TestResourceFastlyFlattenCacheSettings
--- PASS: TestResourceFastlyFlattenCacheSettings (0.00s)
=== RUN   TestAccFastlyServiceV1CacheSetting_basic
--- PASS: TestAccFastlyServiceV1CacheSetting_basic (74.74s)
=== RUN   TestResourceFastlyFlattenConditions
--- PASS: TestResourceFastlyFlattenConditions (0.00s)
=== RUN   TestAccFastlyServiceV1_conditional_basic
--- PASS: TestAccFastlyServiceV1_conditional_basic (41.37s)
=== RUN   TestResourceFastlyFlattenDictionary
--- PASS: TestResourceFastlyFlattenDictionary (0.00s)
=== RUN   TestAccFastlyServiceV1_dictionary
--- PASS: TestAccFastlyServiceV1_dictionary (35.36s)
=== RUN   TestAccFastlyServiceV1_dictionary_update_name
--- PASS: TestAccFastlyServiceV1_dictionary_update_name (78.86s)
=== RUN   TestResourceFastlyFlattenDirectors
--- PASS: TestResourceFastlyFlattenDirectors (0.00s)
=== RUN   TestAccFastlyServiceV1_directors_basic
--- PASS: TestAccFastlyServiceV1_directors_basic (89.73s)
=== RUN   TestResourceFastlyFlattenGCS
--- PASS: TestResourceFastlyFlattenGCS (0.17s)
=== RUN   TestAccFastlyServiceV1_gcslogging
--- PASS: TestAccFastlyServiceV1_gcslogging (36.13s)
=== RUN   TestAccFastlyServiceV1_gcslogging_env
--- PASS: TestAccFastlyServiceV1_gcslogging_env (38.25s)
=== RUN   TestResourceFastlyFlattenGzips
--- PASS: TestResourceFastlyFlattenGzips (0.00s)
=== RUN   TestAccFastlyServiceV1_gzips_basic
--- PASS: TestAccFastlyServiceV1_gzips_basic (77.10s)
=== RUN   TestResourceFastlyFlattenHeaders
--- PASS: TestResourceFastlyFlattenHeaders (0.00s)
=== RUN   TestFastlyServiceV1_BuildHeaders
--- PASS: TestFastlyServiceV1_BuildHeaders (0.00s)
=== RUN   TestAccFastlyServiceV1_headers_basic
--- PASS: TestAccFastlyServiceV1_headers_basic (77.71s)
=== RUN   TestResourceFastlyFlattenHealthChecks
--- PASS: TestResourceFastlyFlattenHealthChecks (0.00s)
=== RUN   TestAccFastlyServiceV1_healthcheck_basic
--- PASS: TestAccFastlyServiceV1_healthcheck_basic (76.24s)
=== RUN   TestResourceFastlyFlattenLogentries
--- PASS: TestResourceFastlyFlattenLogentries (0.00s)
=== RUN   TestAccFastlyServiceV1_logentries_basic
--- PASS: TestAccFastlyServiceV1_logentries_basic (77.73s)
=== RUN   TestAccFastlyServiceV1_logentries_formatVersion
--- PASS: TestAccFastlyServiceV1_logentries_formatVersion (35.80s)
=== RUN   TestResourceFastlyFlattenPapertrail
--- PASS: TestResourceFastlyFlattenPapertrail (0.00s)
=== RUN   TestAccFastlyServiceV1_papertrail_basic
--- PASS: TestAccFastlyServiceV1_papertrail_basic (74.37s)
=== RUN   TestResourceFastlyFlattenRequestSettings
--- PASS: TestResourceFastlyFlattenRequestSettings (0.00s)
=== RUN   TestAccFastlyServiceV1RequestSetting_basic
--- PASS: TestAccFastlyServiceV1RequestSetting_basic (41.27s)
=== RUN   TestResourceFastlyFlattenResponseObjects
--- PASS: TestResourceFastlyFlattenResponseObjects (0.00s)
=== RUN   TestAccFastlyServiceV1_response_object_basic
--- PASS: TestAccFastlyServiceV1_response_object_basic (70.21s)
=== RUN   TestAccFastlyServiceV1_s3logging_basic
--- PASS: TestAccFastlyServiceV1_s3logging_basic (79.29s)
=== RUN   TestAccFastlyServiceV1_s3logging_domain_default
--- PASS: TestAccFastlyServiceV1_s3logging_domain_default (39.00s)
=== RUN   TestAccFastlyServiceV1_s3logging_s3_env
--- PASS: TestAccFastlyServiceV1_s3logging_s3_env (36.30s)
=== RUN   TestAccFastlyServiceV1_s3logging_formatVersion
--- PASS: TestAccFastlyServiceV1_s3logging_formatVersion (36.80s)
=== RUN   TestResourceFastlyFlattenSnippets
--- PASS: TestResourceFastlyFlattenSnippets (0.00s)
=== RUN   TestAccFastlyServiceV1Snippet_basic
--- PASS: TestAccFastlyServiceV1Snippet_basic (70.80s)
=== RUN   TestResourceFastlyFlattenSplunk
--- PASS: TestResourceFastlyFlattenSplunk (0.00s)
=== RUN   TestAccFastlyServiceV1_splunk_basic
--- PASS: TestAccFastlyServiceV1_splunk_basic (77.77s)
=== RUN   TestAccFastlyServiceV1_splunk_default
--- PASS: TestAccFastlyServiceV1_splunk_default (36.91s)
=== RUN   TestAccFastlyServiceV1_splunk_env
--- PASS: TestAccFastlyServiceV1_splunk_env (36.43s)
=== RUN   TestResourceFastlyFlattenSumologic
--- PASS: TestResourceFastlyFlattenSumologic (0.00s)
=== RUN   TestAccFastlyServiceV1_sumologic
--- PASS: TestAccFastlyServiceV1_sumologic (71.37s)
=== RUN   TestResourceFastlyFlattenSyslog
--- PASS: TestResourceFastlyFlattenSyslog (0.00s)
=== RUN   TestAccFastlyServiceV1_syslog_basic
--- PASS: TestAccFastlyServiceV1_syslog_basic (71.07s)
=== RUN   TestAccFastlyServiceV1_syslog_formatVersion
--- PASS: TestAccFastlyServiceV1_syslog_formatVersion (38.30s)
=== RUN   TestResourceFastlyFlattenDomains
--- PASS: TestResourceFastlyFlattenDomains (0.00s)
=== RUN   TestResourceFastlyFlattenBackend
--- PASS: TestResourceFastlyFlattenBackend (0.00s)
=== RUN   TestAccFastlyServiceV1_updateDomain
--- PASS: TestAccFastlyServiceV1_updateDomain (72.50s)
=== RUN   TestAccFastlyServiceV1_updateBackend
--- PASS: TestAccFastlyServiceV1_updateBackend (69.44s)
=== RUN   TestAccFastlyServiceV1_updateInvalidBackend
--- PASS: TestAccFastlyServiceV1_updateInvalidBackend (42.49s)
=== RUN   TestAccFastlyServiceV1_basic
--- PASS: TestAccFastlyServiceV1_basic (80.68s)
=== RUN   TestAccFastlyServiceV1_disappears
--- PASS: TestAccFastlyServiceV1_disappears (18.43s)
=== RUN   TestAccFastlyServiceV1_defaultTTL
--- PASS: TestAccFastlyServiceV1_defaultTTL (116.56s)
=== RUN   TestResourceFastlyFlattenVCLs
--- PASS: TestResourceFastlyFlattenVCLs (0.00s)
=== RUN   TestAccFastlyServiceV1_VCL_basic
--- PASS: TestAccFastlyServiceV1_VCL_basic (78.41s)
=== RUN   TestValidateLoggingFormatVersion
=== RUN   TestValidateLoggingFormatVersion/0
=== RUN   TestValidateLoggingFormatVersion/1
=== RUN   TestValidateLoggingFormatVersion/2
=== RUN   TestValidateLoggingFormatVersion/3
=== RUN   TestValidateLoggingFormatVersion/4
=== RUN   TestValidateLoggingFormatVersion/5
--- PASS: TestValidateLoggingFormatVersion (0.00s)
    --- PASS: TestValidateLoggingFormatVersion/0 (0.00s)
    --- PASS: TestValidateLoggingFormatVersion/1 (0.00s)
    --- PASS: TestValidateLoggingFormatVersion/2 (0.00s)
    --- PASS: TestValidateLoggingFormatVersion/3 (0.00s)
    --- PASS: TestValidateLoggingFormatVersion/4 (0.00s)
    --- PASS: TestValidateLoggingFormatVersion/5 (0.00s)
=== RUN   TestValidateLoggingMessageType
=== RUN   TestValidateLoggingMessageType/classic
=== RUN   TestValidateLoggingMessageType/loggly
=== RUN   TestValidateLoggingMessageType/logplex
=== RUN   TestValidateLoggingMessageType/blank
=== RUN   TestValidateLoggingMessageType/CLASSIC
=== RUN   TestValidateLoggingMessageType/LOGGLY
=== RUN   TestValidateLoggingMessageType/LOGPLEX
=== RUN   TestValidateLoggingMessageType/BLANK
--- PASS: TestValidateLoggingMessageType (0.00s)
    --- PASS: TestValidateLoggingMessageType/classic (0.00s)
    --- PASS: TestValidateLoggingMessageType/loggly (0.00s)
    --- PASS: TestValidateLoggingMessageType/logplex (0.00s)
    --- PASS: TestValidateLoggingMessageType/blank (0.00s)
    --- PASS: TestValidateLoggingMessageType/CLASSIC (0.00s)
    --- PASS: TestValidateLoggingMessageType/LOGGLY (0.00s)
    --- PASS: TestValidateLoggingMessageType/LOGPLEX (0.00s)
    --- PASS: TestValidateLoggingMessageType/BLANK (0.00s)
=== RUN   TestValidateLoggingPlacement
=== RUN   TestValidateLoggingPlacement/none
=== RUN   TestValidateLoggingPlacement/waf_debug
=== RUN   TestValidateLoggingPlacement/NONE
=== RUN   TestValidateLoggingPlacement/WAF_DEBUG
--- PASS: TestValidateLoggingPlacement (0.00s)
    --- PASS: TestValidateLoggingPlacement/none (0.00s)
    --- PASS: TestValidateLoggingPlacement/waf_debug (0.00s)
    --- PASS: TestValidateLoggingPlacement/NONE (0.00s)
    --- PASS: TestValidateLoggingPlacement/WAF_DEBUG (0.00s)
=== RUN   TestValidateDirectorQuorum
=== RUN   TestValidateDirectorQuorum/0
=== RUN   TestValidateDirectorQuorum/55
=== RUN   TestValidateDirectorQuorum/100
=== RUN   TestValidateDirectorQuorum/-1
=== RUN   TestValidateDirectorQuorum/101
=== RUN   TestValidateDirectorQuorum/150
--- PASS: TestValidateDirectorQuorum (0.00s)
    --- PASS: TestValidateDirectorQuorum/0 (0.00s)
    --- PASS: TestValidateDirectorQuorum/55 (0.00s)
    --- PASS: TestValidateDirectorQuorum/100 (0.00s)
    --- PASS: TestValidateDirectorQuorum/-1 (0.00s)
    --- PASS: TestValidateDirectorQuorum/101 (0.00s)
    --- PASS: TestValidateDirectorQuorum/150 (0.00s)
=== RUN   TestValidateDirectorType
=== RUN   TestValidateDirectorType/5
=== RUN   TestValidateDirectorType/0
=== RUN   TestValidateDirectorType/1
=== RUN   TestValidateDirectorType/2
=== RUN   TestValidateDirectorType/3
=== RUN   TestValidateDirectorType/4
--- PASS: TestValidateDirectorType (0.00s)
    --- PASS: TestValidateDirectorType/5 (0.00s)
    --- PASS: TestValidateDirectorType/0 (0.00s)
    --- PASS: TestValidateDirectorType/1 (0.00s)
    --- PASS: TestValidateDirectorType/2 (0.00s)
    --- PASS: TestValidateDirectorType/3 (0.00s)
    --- PASS: TestValidateDirectorType/4 (0.00s)
=== RUN   TestValidateConditionType
=== RUN   TestValidateConditionType/REQUEST
=== RUN   TestValidateConditionType/RESPONSE
=== RUN   TestValidateConditionType/CACHE
=== RUN   TestValidateConditionType/request
=== RUN   TestValidateConditionType/response
=== RUN   TestValidateConditionType/cache
--- PASS: TestValidateConditionType (0.00s)
    --- PASS: TestValidateConditionType/REQUEST (0.00s)
    --- PASS: TestValidateConditionType/RESPONSE (0.00s)
    --- PASS: TestValidateConditionType/CACHE (0.00s)
    --- PASS: TestValidateConditionType/request (0.00s)
    --- PASS: TestValidateConditionType/response (0.00s)
    --- PASS: TestValidateConditionType/cache (0.00s)
=== RUN   TestValidateHeaderAction
=== RUN   TestValidateHeaderAction/set
=== RUN   TestValidateHeaderAction/append
=== RUN   TestValidateHeaderAction/delete
=== RUN   TestValidateHeaderAction/regex
=== RUN   TestValidateHeaderAction/regex_repeat
=== RUN   TestValidateHeaderAction/SET
=== RUN   TestValidateHeaderAction/APPEND
=== RUN   TestValidateHeaderAction/DELETE
=== RUN   TestValidateHeaderAction/REGEX
=== RUN   TestValidateHeaderAction/REGEX_REPEAT
--- PASS: TestValidateHeaderAction (0.00s)
    --- PASS: TestValidateHeaderAction/set (0.00s)
    --- PASS: TestValidateHeaderAction/append (0.00s)
    --- PASS: TestValidateHeaderAction/delete (0.00s)
    --- PASS: TestValidateHeaderAction/regex (0.00s)
    --- PASS: TestValidateHeaderAction/regex_repeat (0.00s)
    --- PASS: TestValidateHeaderAction/SET (0.00s)
    --- PASS: TestValidateHeaderAction/APPEND (0.00s)
    --- PASS: TestValidateHeaderAction/DELETE (0.00s)
    --- PASS: TestValidateHeaderAction/REGEX (0.00s)
    --- PASS: TestValidateHeaderAction/REGEX_REPEAT (0.00s)
=== RUN   TestValidateHeaderType
=== RUN   TestValidateHeaderType/request
=== RUN   TestValidateHeaderType/fetch
=== RUN   TestValidateHeaderType/cache
=== RUN   TestValidateHeaderType/response
=== RUN   TestValidateHeaderType/REQUEST
=== RUN   TestValidateHeaderType/FETCH
=== RUN   TestValidateHeaderType/CACHE
=== RUN   TestValidateHeaderType/RESPONSE
--- PASS: TestValidateHeaderType (0.00s)
    --- PASS: TestValidateHeaderType/request (0.00s)
    --- PASS: TestValidateHeaderType/fetch (0.00s)
    --- PASS: TestValidateHeaderType/cache (0.00s)
    --- PASS: TestValidateHeaderType/response (0.00s)
    --- PASS: TestValidateHeaderType/REQUEST (0.00s)
    --- PASS: TestValidateHeaderType/FETCH (0.00s)
    --- PASS: TestValidateHeaderType/CACHE (0.00s)
    --- PASS: TestValidateHeaderType/RESPONSE (0.00s)
=== RUN   TestValidateSnippetType
=== RUN   TestValidateSnippetType/init
=== RUN   TestValidateSnippetType/recv
=== RUN   TestValidateSnippetType/hit
=== RUN   TestValidateSnippetType/miss
=== RUN   TestValidateSnippetType/pass
=== RUN   TestValidateSnippetType/fetch
=== RUN   TestValidateSnippetType/error
=== RUN   TestValidateSnippetType/deliver
=== RUN   TestValidateSnippetType/log
=== RUN   TestValidateSnippetType/none
=== RUN   TestValidateSnippetType/INIT
=== RUN   TestValidateSnippetType/RECV
=== RUN   TestValidateSnippetType/HIT
=== RUN   TestValidateSnippetType/MISS
=== RUN   TestValidateSnippetType/PASS
=== RUN   TestValidateSnippetType/FETCH
=== RUN   TestValidateSnippetType/ERROR
=== RUN   TestValidateSnippetType/DELIVER
=== RUN   TestValidateSnippetType/LOG
=== RUN   TestValidateSnippetType/NONE
--- PASS: TestValidateSnippetType (0.00s)
    --- PASS: TestValidateSnippetType/init (0.00s)
    --- PASS: TestValidateSnippetType/recv (0.00s)
    --- PASS: TestValidateSnippetType/hit (0.00s)
    --- PASS: TestValidateSnippetType/miss (0.00s)
    --- PASS: TestValidateSnippetType/pass (0.00s)
    --- PASS: TestValidateSnippetType/fetch (0.00s)
    --- PASS: TestValidateSnippetType/error (0.00s)
    --- PASS: TestValidateSnippetType/deliver (0.00s)
    --- PASS: TestValidateSnippetType/log (0.00s)
    --- PASS: TestValidateSnippetType/none (0.00s)
    --- PASS: TestValidateSnippetType/INIT (0.00s)
    --- PASS: TestValidateSnippetType/RECV (0.00s)
    --- PASS: TestValidateSnippetType/HIT (0.00s)
    --- PASS: TestValidateSnippetType/MISS (0.00s)
    --- PASS: TestValidateSnippetType/PASS (0.00s)
    --- PASS: TestValidateSnippetType/FETCH (0.00s)
    --- PASS: TestValidateSnippetType/ERROR (0.00s)
    --- PASS: TestValidateSnippetType/DELIVER (0.00s)
    --- PASS: TestValidateSnippetType/LOG (0.00s)
    --- PASS: TestValidateSnippetType/NONE (0.00s)
PASS
ok      github.com/terraform-providers/terraform-provider-fastly/fastly 2569.047s
?       github.com/terraform-providers/terraform-provider-fastly/version        [no test files]
```

</p>
</details>

phamann commented on fastly/resource_fastly_service_dictionary_items_v1.go:
```diff
@@ -0,0 +1,219 @@
+	items := d.Get("items").(map[string]interface{})
+
+	for k, v := range items {
+		_, err := conn.CreateDictionaryItem(&gofastly.CreateDictionaryItemInput{
```
Instead of looping over the items and calling `CreateDictionaryItem` for each item, it would be more efficient to use the [batch API endpoint](https://docs.fastly.com/api/config#dictionary_item_dc826ce1255a7c42bc48eb204eed8f7f) and use [`CreateDictionaryItems`](https://godoc.org/github.com/fastly/go-fastly/fastly#Client.CreateDictionaryItems) instead.

trentrosenbaum replied:
The Create, Update and Delete functions now use batch operations to alter the fastly remote state.  The Read function still uses the list function.

phamann commented on fastly/resource_fastly_service_dictionary_items_v1.go:
```diff
@@ -0,0 +1,219 @@
+	}
+
+	// TODO review this
+	d.SetId(fmt.Sprintf("%s/%s", serviceID, dictionaryID))
```
As per the above `TODO`; what needs to be reviewed? Whether this is an acceptable ID?

trentrosenbaum replied:
This was concerning the structure of the ID.  The serviceID and dictionaryID are not being derived from this value and so the resource ID is not being sent to the remote state.  I have removed the TODO

phamann commented on website/docs/r/service_dictionary_items_v1.html.markdown:
```diff
@@ -0,0 +1,156 @@
+  Provides a grouping of fastly dictionary items that can be applied to a service. 
+---
+
+# fastly_service_dictionary_items_v1
```
It would be good to add a comment in this documentation regarding the use of lifecycle ignore_changes if the user doesn't want TF to manage item state after initial apply.

trentrosenbaum replied:
I have expanded the description to cover this and provided an example of the lifecycle ignore_changes with the resource.

trentrosenbaum replied:
documentation has been expanded and will need a more specific review.

phamann commented on fastly/resource_fastly_service_dictionary_items_v1.go:
```diff
@@ -0,0 +1,219 @@
+		// Handle Removal
+		for k := range os {
+			if _, ok := ns[k]; !ok {
+				err := conn.DeleteDictionaryItem(&gofastly.DeleteDictionaryItemInput{
```
As per my above comment, it would be better if this could use the batch endpoint. However it seems that the `go-fastly` client lacks a `DeleteDictionaryItems` method. 😢

trentrosenbaum replied:
This function now uses the batch operation provided by the go-fastly client.

phamann commented on fastly/resource_fastly_service_dictionary_items_v1.go:
```diff
@@ -0,0 +1,219 @@
+			// Handle replaces
+			if _, ok := os[k]; ok {
+
+				_, err := conn.UpdateDictionaryItem(&gofastly.UpdateDictionaryItemInput{
```
As per my above comment, it would be better if this could use the batch endpoint. However it seems that the `go-fastly` client lacks a `UpdateDictionaryItems` method. 😢

trentrosenbaum replied:
This function now uses the batch operations provided by the go-fastly client.

phamann commented on fastly/resource_fastly_service_dictionary_items_v1.go:
```diff
@@ -0,0 +1,219 @@
+			// Handle additions
+			if _, ok := os[k]; !ok {
+
+				_, err := conn.CreateDictionaryItem(&gofastly.CreateDictionaryItemInput{
```
As per above, we should use the `CreateDictionaryItems` method here too.

trentrosenbaum replied:
This function now uses the batch operations provided by the go-fastly client.

phamann reviewed (commented):
Looks good overall, however we should discuss the API endpoint  / client method usage. 

Until we have the correct methods exposed in the client `UpdateDictionaryItems`, `DeleteDictionaryItems` etc, you could construct the API call manually and use the raw `PatchJSON()` method, but seems like a nasty hack.

Commit:
Updates documentation to explain using lifecycle ignore_changes with regards to dictionary items.

Commit:
Removed TODO concerning resource id.

Commit:
Updates the repo to use the implement-batch-operations-for-dictionary-items of the go-fastly client.

Commit:
Updated to use refactored go-fastly dictionary batch

Commit:
Updates the dictionary items resource by removing filter and applies validation function for max number of items.

matpimenta commented:
@trentrosenbaum Travis is failing, you will need to run `make fmt`

Commit:
Corrected formating of dictionary items resource and validators

Commit:
Corrects the validation for dictionary items

matpimenta commented on fastly/resource_fastly_service_dictionary_items_v1_test.go:
```diff
@@ -0,0 +1,483 @@
+				Config:    config,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckServiceV1Exists("fastly_service_v1.foo", &service),
+					testAccCheckFastlyServiceDictionaryItemsV1RemoteState(&service, name, dictName, expectedRemoteItemsAfterUpdate),
```
The expected value should be `expectedRemoteItems` at this point

trentrosenbaum replied:
The tests have been adjusted to reflect the correct expectations.

matpimenta commented on fastly/resource_fastly_service_dictionary_items_v1_test.go:
```diff
@@ -0,0 +1,483 @@
+	})
+}
+
+func TestAccFastlyServiceDictionaryItemV1_external_item_is_ignored(t *testing.T) {
```
Maybe `TestAccFastlyServiceDictionaryItemV1_external_item_is_removed` is a better description for the new behaviour

trentrosenbaum replied:
The tests have been adjusted to reflect the correct expectations.

Commit:
Updates the resource website page with initial wording describing expectations.

matpimenta committed:
Adding import function and related test / Refactor dictionary item tests to reduce duplication

Commit:
Updates the dictionary items test to ensure local and remotely add items are deleted.

Commit:
Updates to use go-fastly v1.2.0

trentrosenbaum committed:
Merge pull request #1 from opencredo/dict-resource-import

Dict resource import

Commit:
Reinstates the items ValidateFunc

Commit:
Updates the website documentation to cover imports of dictionary items.

phamann commented on fastly/resource_fastly_service_dictionary_items_v1.go:
```diff
@@ -0,0 +1,239 @@
+			"items": {
+				Type:         schema.TypeMap,
+				Optional:     true,
+				Description:  "Key/value pairs that make up an item in the dictionary",
```
I suggest prefixing this with "Map of ", ie:
"Map of key/value pairs that make up an item in the dictionary"

trentrosenbaum replied:
Message was updated with new wording.

phamann commented on fastly/resource_fastly_service_v1.go:
```diff
@@ -1385,6 +1385,26 @@ func resourceServiceV1() *schema.Resource {
+						"dictionary_id": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "Generated dictionary id",
```
Change `id` to `ID`

trentrosenbaum replied:
Files were searched and updated where appropriate.

phamann commented on fastly/validators.go:
```diff
@@ -74,3 +75,24 @@ func validateSnippetType() schema.SchemaValidateFunc {
 }
+
+func validateDictionaryItems() schema.SchemaValidateFunc {
+	max := 10000
```
Can we store this is a package level constant like we did with `gofastly.BatchModifyMaximumOperations`.

trentrosenbaum replied:
So we could place it as a value in the go-fastly client package.  This way at least it is documented more towards the API level.  It can then be consumed at the Terraform provider validation level.

I can put in two constants.  The first will represent the max number of dictionary items and the second will represent the max for acl entries.  Thinking at these two values can be set independently.

phamann replied:
Sounds sane to me! :+1:

trentrosenbaum replied:
The validation function now uses the planned constant from the go-fastly client v1.2.1

phamann commented on website/docs/r/service_dictionary_items_v1.html.markdown:
```diff
@@ -0,0 +1,212 @@
+page_title: "Fastly: service_dictionary_items_v1"
+sidebar_current: "docs-fastly-resource-service-v1"
+description: |-
+  Provides a grouping of fastly dictionary items that can be applied to a service. 
```
Change to:
> Provides a map of Fastly dictionary items that can be applied to a service.

trentrosenbaum replied:
Message was updated with new wording.

phamann commented on website/docs/r/service_dictionary_items_v1.html.markdown:
```diff
@@ -0,0 +1,212 @@
+Provides a grouping of fastly dictionary items that can be applied to a service.
+ 
+This resource will populate a dictionary with the items and it will track their state going forward.
+Additional dictionary items can be added through the Fastly API/console, but will be removed from fastly if terraform detects a difference in the remote state.
```
Please change "API/console" to "API or UI"
Please change "faslty" to "Fastly"
Please change "terraform" to "Terraform"

trentrosenbaum replied:
Files were searched and updated where appropriate.

phamann commented on website/docs/r/service_dictionary_items_v1.html.markdown:
```diff
@@ -0,0 +1,212 @@
+This resource will populate a dictionary with the items and it will track their state going forward.
+Additional dictionary items can be added through the Fastly API/console, but will be removed from fastly if terraform detects a difference in the remote state.
+
+The Fastly API/console can also be used to modify the items that are managed through Terraform.  In this case the default behaviour of the 
```
Please change "API/console" to "API or UI"

trentrosenbaum replied:
Files were searched and updated where appropriate.

phamann commented on website/docs/r/service_dictionary_items_v1.html.markdown:
```diff
@@ -0,0 +1,212 @@
+Additional dictionary items can be added through the Fastly API/console, but will be removed from fastly if terraform detects a difference in the remote state.
+
+The Fastly API/console can also be used to modify the items that are managed through Terraform.  In this case the default behaviour of the 
+resource will be to realign the remoted state.  The items in the remote fastly dictionary will be updated or deleted according to the terraform plan.  
```
Please change "fastly" to "Fastly"

trentrosenbaum replied:
Files were searched and updated where appropriate.

phamann commented on website/docs/r/service_dictionary_items_v1.html.markdown:
```diff
@@ -0,0 +1,212 @@
+The Fastly API/console can also be used to modify the items that are managed through Terraform.  In this case the default behaviour of the 
+resource will be to realign the remoted state.  The items in the remote fastly dictionary will be updated or deleted according to the terraform plan.  
+
+If Terraform is being used to only create initial items in a dictionary, then the lifecyle ignore_changes field can be used with the resource.  An example of this configuraiton is provided below.    
```
Please wrap "ignore_changes" with backticks so it is rendered appropriately.

trentrosenbaum replied:
The extra formating was added to render correctly

phamann commented on website/docs/r/service_dictionary_items_v1.html.markdown:
```diff
@@ -0,0 +1,212 @@
+
+Lifecyle ignore_changes usage:
+
+The following example demonstrates how the lifecycle ignore_change field can be used to suppress updates against the 
```
Please wrap `ignore_changes` in backticks so it is rendered appropriately.

trentrosenbaum replied:
Files were searched and updated where appropriate.

phamann commented on website/docs/r/service_dictionary_items_v1.html.markdown:
```diff
@@ -0,0 +1,212 @@
+Lifecyle ignore_changes usage:
+
+The following example demonstrates how the lifecycle ignore_change field can be used to suppress updates against the 
+items in a dictionary.  If an external means, (for example via the Fastly API) is used to manage items in a dictionary, then this will stop terraform realigning the remote state.
```
Please change "terraform" to "Terraform"

trentrosenbaum replied:
Files were searched and updated where appropriate.

phamann commented on website/docs/r/service_dictionary_items_v1.html.markdown:
```diff
@@ -0,0 +1,212 @@
+
+## Import
+
+Dictionary items can be populated through the Fastly API and/or the console.  These items can then be managed by terraform through the use of the import command.
```
Please change "console" to "UI"
Please change "terraform" to "Terraform"

trentrosenbaum replied:
Files were searched and updated where appropriate.

phamann commented on website/docs/r/service_dictionary_items_v1.html.markdown:
```diff
@@ -0,0 +1,212 @@
+## Import
+
+Dictionary items can be populated through the Fastly API and/or the console.  These items can then be managed by terraform through the use of the import command.
+Resources are imported into the terraform state through the use of an ID, and the fastly_service_dictionary_items_v1 resouce has an ID that is made up of the following format.
```
Please change "terraform" to "Terraform"

trentrosenbaum replied:
Files were searched and updated where appropriate.

phamann commented on website/docs/r/service_dictionary_items_v1.html.markdown:
````diff
@@ -0,0 +1,212 @@
+$ terraform import fastly_service_dictionary_items_v1.items <serviceID>/<dictionaryID>
+```
+
+If terraform is already managing remote dictionary items against a resource being imported then the user will be asked to remove it from the existing terraform state.  
````
Please change "terraform" to "Terraform"

phamann commented on website/docs/r/service_dictionary_items_v1.html.markdown:
````diff
@@ -0,0 +1,212 @@
+```
+
+If terraform is already managing remote dictionary items against a resource being imported then the user will be asked to remove it from the existing terraform state.  
+The following is an example of the terraform state command to remove the resource named `fastly_service_dictionary_items_v1.items` from the terraform state file.
````
Please change "terraform" to "Terraform"

trentrosenbaum replied:
Files were searched and updated where appropriate.

Commit:
Adjusts the documentation as requested in the documentation review.

triblondon commented on website/docs/r/service_dictionary_items_v1.html.markdown:
```diff
@@ -0,0 +1,212 @@
+
+# fastly_service_dictionary_items_v1
+
+Provides a map of Fastly dictionary items that can be applied to a service.
```
"Provides"? Maybe "Defines"?

triblondon replied:
"can be applied to a service" => "can be used to populate a service dictionary"?

trentrosenbaum replied:
Alterations have been applied

triblondon commented on website/docs/r/service_dictionary_items_v1.html.markdown:
```diff
@@ -0,0 +1,212 @@
+
+Provides a map of Fastly dictionary items that can be applied to a service.
+ 
+This resource will populate a dictionary with the items and it will track their state going forward.
```
```suggestion
This resource will populate a dictionary with the items and will track their state.
```

triblondon commented on website/docs/r/service_dictionary_items_v1.html.markdown:
```diff
@@ -0,0 +1,212 @@
+Provides a map of Fastly dictionary items that can be applied to a service.
+ 
+This resource will populate a dictionary with the items and it will track their state going forward.
+Additional dictionary items can be added through the Fastly API or UI, but will be removed from Fastly if Terraform detects a difference in the remote state.
```
```suggestion
Dictionary items can also be added through the Fastly API or UI, but will be removed from Fastly if Terraform detects a difference in the remote state.
```

Can this be in a warning box?

triblondon commented on website/docs/r/service_dictionary_items_v1.html.markdown:
```diff
@@ -0,0 +1,212 @@
+Additional dictionary items can be added through the Fastly API or UI, but will be removed from Fastly if Terraform detects a difference in the remote state.
+
+The Fastly API or UI can also be used to modify the items that are managed through Terraform.  In this case the default behaviour of the 
+resource will be to realign the remoted state.  The items in the remote Fastly dictionary will be updated or deleted according to the Terraform plan.  
```
This seems a bit complicated way of saying "Terraform will take precedence over any changes you make in the UI or API.  Such changes are likely to be reversed if you run terraform again.

trentrosenbaum replied:
Paragraphs have been replaced by the more straightforward wording

triblondon commented on website/docs/r/service_dictionary_items_v1.html.markdown:
```diff
@@ -0,0 +1,212 @@
+The Fastly API or UI can also be used to modify the items that are managed through Terraform.  In this case the default behaviour of the 
+resource will be to realign the remoted state.  The items in the remote Fastly dictionary will be updated or deleted according to the Terraform plan.  
+
+If Terraform is being used to only create initial items in a dictionary, then the lifecyle `ignore_changes` field can be used with the resource.  An example of this configuraiton is provided below.    
```
```suggestion
If Terraform is being used to populate the initial content of a dictionary which you intend to manage via API or UI, then the lifecycle `ignore_changes` field can be used with the resource.  An example of this configuration is provided below.    
```

triblondon commented on website/docs/r/service_dictionary_items_v1.html.markdown:
````diff
@@ -0,0 +1,212 @@
+}
+```
+
+Lifecyle ignore_changes usage:
````
Make a heading?

Rename to "Supporting API and UI dictionary updates with ignore_changes"

trentrosenbaum replied:
Changed to heading and text changed to suggested.

triblondon commented on website/docs/r/service_dictionary_items_v1.html.markdown:
```diff
@@ -0,0 +1,212 @@
+Lifecyle ignore_changes usage:
+
+The following example demonstrates how the lifecycle ignore_change field can be used to suppress updates against the 
+items in a dictionary.  If an external means, (for example via the Fastly API or UI) is used to manage items in a dictionary, then this will stop Terraform realigning the remote state.
```
```suggestion
items in a dictionary.  If, after your first deploy, the Fastly API or UI is to be used to manage items in a dictionary, then this will stop Terraform realigning the remote state with the initial set of dictionary items defined in your HCL.
```

triblondon commented on website/docs/r/service_dictionary_items_v1.html.markdown:
```diff
@@ -0,0 +1,212 @@
+
+The following arguments are supported:
+
+* `service_id` - (Required) The ID of the Service that the dictionary belongs to
```
Why cap on Service?

triblondon replied:
If it's a resource, backtick it

triblondon commented on website/docs/r/service_dictionary_items_v1.html.markdown:
```diff
@@ -0,0 +1,212 @@
+
+* `service_id` - (Required) The ID of the Service that the dictionary belongs to
+* `dictionary_id` - (Required) The ID of the dictionary that the items belong to
+* `items` - (Optional) A Map representing an entry in the dictionary, (key/value)
```
Ditto on Map

triblondon commented on website/docs/r/service_dictionary_items_v1.html.markdown:
```diff
@@ -0,0 +1,212 @@
+## Attributes Reference
+
+[fastly-dictionary]: https://docs.fastly.com/api/config#dictionary
+[fastly-dictionary_item]: https://docs.fastly.com/api/config#dictionary_item
```
Is this valid markdown? 🤔

trentrosenbaum replied:
This was correct.  The markdown was not correct.  The links have been updated and now display.

triblondon commented on website/docs/r/service_dictionary_items_v1.html.markdown:
```diff
@@ -0,0 +1,212 @@
+
+## Import
+
+Dictionary items can be populated through the Fastly API or UI.  These items can then be managed by Terraform through the use of the import command.
```
```suggestion
Dictionary items in your HCL can be populated by pulling the current remote state of an existing Fastly service, using the `import` command.
```

triblondon commented on website/docs/r/service_dictionary_items_v1.html.markdown:
```diff
@@ -0,0 +1,212 @@
+## Import
+
+Dictionary items can be populated through the Fastly API or UI.  These items can then be managed by Terraform through the use of the import command.
+Resources are imported into the Terraform state through the use of an ID, and the fastly_service_dictionary_items_v1 resouce has an ID that is made up of the following format.
```
I would delete this and the subsequent code block.

trentrosenbaum replied:
The wording and code have been removed.

triblondon commented on website/docs/r/service_dictionary_items_v1.html.markdown:
````diff
@@ -0,0 +1,212 @@
+This is an example of the import command being applied to the resource named `fastly_service_dictionary_items_v1.items`
+
+```
+$ terraform import fastly_service_dictionary_items_v1.items <serviceID>/<dictionaryID>
````
I would rewrite this to be either purely a syntactic guide or purely an example, not a combination of both.

trentrosenbaum replied:
The example has been chosen and values used in the place holders.

Commit:
Udpates to use go-fastly v1.2.1 and adjust dictionary items validation.

trentrosenbaum committed:
Update website/docs/r/service_dictionary_items_v1.html.markdown

Co-Authored-By: Andrew Betts <andrew@trib.tv>

trentrosenbaum committed:
Update website/docs/r/service_dictionary_items_v1.html.markdown

Co-Authored-By: Andrew Betts <andrew@trib.tv>

trentrosenbaum committed:
Update website/docs/r/service_dictionary_items_v1.html.markdown

Co-Authored-By: Andrew Betts <andrew@trib.tv>

trentrosenbaum committed:
Update website/docs/r/service_dictionary_items_v1.html.markdown

Co-Authored-By: Andrew Betts <andrew@trib.tv>

Commit:
Udpated to include documentation feedback.

trentrosenbaum committed:
Update website/docs/r/service_dictionary_items_v1.html.markdown

Co-Authored-By: Andrew Betts <andrew@trib.tv>

Commit:
Udpated to include documentation feedback.

Commit:
Merge branch 'upgrade-provider-to-0.12.x-with-dict-resource' of github.com:opencredo/terraform-provider-fastly into upgrade-provider-to-0.12.x-with-dict-resource

Commit:
Updated to include documentation feedback.

Commit:
Updated to include documentation feedback.

Commit:
Updated with warning marker to highlight terraform precedence

Commit:
Adds markdown characters to render coloured area around warning text.

trentrosenbaum commented:
The following shows the result of the acceptance test execution against this branch/PR
The import has been added and the documentation changes adjusted with feedback.


<details><summary>Acceptance Test Results</summary>
<p>

```bash
$ make testacc
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v  -timeout 120m -ldflags="-X=github.com/terraform-providers/terraform-provider-fastly/version.ProviderVersion=acc"
?       github.com/terraform-providers/terraform-provider-fastly        [no test files]
=== RUN   TestUserAgentContainsProviderVersion
--- PASS: TestUserAgentContainsProviderVersion (0.00s)
=== RUN   TestAccFastlyIPRanges
--- PASS: TestAccFastlyIPRanges (0.92s)
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestResourceFastlyFlattenDictionaryItems
--- PASS: TestResourceFastlyFlattenDictionaryItems (0.00s)
=== RUN   TestAccFastlyServiceDictionaryItemV1_create
--- PASS: TestAccFastlyServiceDictionaryItemV1_create (72.79s)
=== RUN   TestAccFastlyServiceDictionaryItemV1_create_dynamic
--- PASS: TestAccFastlyServiceDictionaryItemV1_create_dynamic (71.15s)
=== RUN   TestAccFastlyServiceDictionaryItemV1_update
--- PASS: TestAccFastlyServiceDictionaryItemV1_update (143.90s)
=== RUN   TestAccFastlyServiceDictionaryItemV1_external_item_is_removed
--- PASS: TestAccFastlyServiceDictionaryItemV1_external_item_is_removed (103.69s)
=== RUN   TestAccFastlyServiceDictionaryItemV1_external_item_deleted
--- PASS: TestAccFastlyServiceDictionaryItemV1_external_item_deleted (151.19s)
=== RUN   TestAccFastlyServiceDictionaryItemV1_batch_1001_items
--- PASS: TestAccFastlyServiceDictionaryItemV1_batch_1001_items (120.57s)
=== RUN   TestAccFastlyServiceDictionaryItemV1_import
--- PASS: TestAccFastlyServiceDictionaryItemV1_import (68.53s)
=== RUN   TestResourceFastlyFlattenBigQuery
--- PASS: TestResourceFastlyFlattenBigQuery (0.51s)
=== RUN   TestAccFastlyServiceV1_bigquerylogging
--- PASS: TestAccFastlyServiceV1_bigquerylogging (67.72s)
=== RUN   TestAccFastlyServiceV1_bigquerylogging_env
--- PASS: TestAccFastlyServiceV1_bigquerylogging_env (71.18s)
=== RUN   TestResourceFastlyFlattenBlobStorage
--- PASS: TestResourceFastlyFlattenBlobStorage (0.00s)
=== RUN   TestAccFastlyServiceV1_blobstoragelogging_basic
--- PASS: TestAccFastlyServiceV1_blobstoragelogging_basic (139.45s)
=== RUN   TestAccFastlyServiceV1_blobstoragelogging_default
--- PASS: TestAccFastlyServiceV1_blobstoragelogging_default (65.99s)
=== RUN   TestAccFastlyServiceV1_blobstoragelogging_env
--- PASS: TestAccFastlyServiceV1_blobstoragelogging_env (65.52s)
=== RUN   TestResourceFastlyFlattenCacheSettings
--- PASS: TestResourceFastlyFlattenCacheSettings (0.00s)
=== RUN   TestAccFastlyServiceV1CacheSetting_basic
--- PASS: TestAccFastlyServiceV1CacheSetting_basic (131.44s)
=== RUN   TestResourceFastlyFlattenConditions
--- PASS: TestResourceFastlyFlattenConditions (0.00s)
=== RUN   TestAccFastlyServiceV1_conditional_basic
--- PASS: TestAccFastlyServiceV1_conditional_basic (65.88s)
=== RUN   TestResourceFastlyFlattenDictionary
--- PASS: TestResourceFastlyFlattenDictionary (0.00s)
=== RUN   TestAccFastlyServiceV1_dictionary
--- PASS: TestAccFastlyServiceV1_dictionary (59.82s)
=== RUN   TestAccFastlyServiceV1_dictionary_update_name
--- PASS: TestAccFastlyServiceV1_dictionary_update_name (133.02s)
=== RUN   TestResourceFastlyFlattenDirectors
--- PASS: TestResourceFastlyFlattenDirectors (0.00s)
=== RUN   TestAccFastlyServiceV1_directors_basic
--- PASS: TestAccFastlyServiceV1_directors_basic (152.16s)
=== RUN   TestResourceFastlyFlattenGCS
--- PASS: TestResourceFastlyFlattenGCS (0.48s)
=== RUN   TestAccFastlyServiceV1_gcslogging
--- PASS: TestAccFastlyServiceV1_gcslogging (73.06s)
=== RUN   TestAccFastlyServiceV1_gcslogging_env
--- PASS: TestAccFastlyServiceV1_gcslogging_env (64.93s)
=== RUN   TestResourceFastlyFlattenGzips
--- PASS: TestResourceFastlyFlattenGzips (0.00s)
=== RUN   TestAccFastlyServiceV1_gzips_basic
--- PASS: TestAccFastlyServiceV1_gzips_basic (132.61s)
=== RUN   TestResourceFastlyFlattenHeaders
--- PASS: TestResourceFastlyFlattenHeaders (0.00s)
=== RUN   TestFastlyServiceV1_BuildHeaders
--- PASS: TestFastlyServiceV1_BuildHeaders (0.00s)
=== RUN   TestAccFastlyServiceV1_headers_basic
--- PASS: TestAccFastlyServiceV1_headers_basic (126.98s)
=== RUN   TestResourceFastlyFlattenHealthChecks
--- PASS: TestResourceFastlyFlattenHealthChecks (0.00s)
=== RUN   TestAccFastlyServiceV1_healthcheck_basic
--- PASS: TestAccFastlyServiceV1_healthcheck_basic (131.49s)
=== RUN   TestResourceFastlyFlattenLogentries
--- PASS: TestResourceFastlyFlattenLogentries (0.00s)
=== RUN   TestAccFastlyServiceV1_logentries_basic
--- PASS: TestAccFastlyServiceV1_logentries_basic (131.26s)
=== RUN   TestAccFastlyServiceV1_logentries_formatVersion
--- PASS: TestAccFastlyServiceV1_logentries_formatVersion (63.81s)
=== RUN   TestResourceFastlyFlattenPapertrail
--- PASS: TestResourceFastlyFlattenPapertrail (0.00s)
=== RUN   TestAccFastlyServiceV1_papertrail_basic
--- PASS: TestAccFastlyServiceV1_papertrail_basic (126.88s)
=== RUN   TestResourceFastlyFlattenRequestSettings
--- PASS: TestResourceFastlyFlattenRequestSettings (0.00s)
=== RUN   TestAccFastlyServiceV1RequestSetting_basic
--- PASS: TestAccFastlyServiceV1RequestSetting_basic (69.93s)
=== RUN   TestResourceFastlyFlattenResponseObjects
--- PASS: TestResourceFastlyFlattenResponseObjects (0.00s)
=== RUN   TestAccFastlyServiceV1_response_object_basic
--- PASS: TestAccFastlyServiceV1_response_object_basic (128.89s)
=== RUN   TestAccFastlyServiceV1_s3logging_basic
--- PASS: TestAccFastlyServiceV1_s3logging_basic (132.35s)
=== RUN   TestAccFastlyServiceV1_s3logging_domain_default
--- PASS: TestAccFastlyServiceV1_s3logging_domain_default (71.89s)
=== RUN   TestAccFastlyServiceV1_s3logging_s3_env
--- PASS: TestAccFastlyServiceV1_s3logging_s3_env (59.93s)
=== RUN   TestAccFastlyServiceV1_s3logging_formatVersion
--- PASS: TestAccFastlyServiceV1_s3logging_formatVersion (58.35s)
=== RUN   TestResourceFastlyFlattenSnippets
--- PASS: TestResourceFastlyFlattenSnippets (0.00s)
=== RUN   TestAccFastlyServiceV1Snippet_basic
--- PASS: TestAccFastlyServiceV1Snippet_basic (128.00s)
=== RUN   TestResourceFastlyFlattenSplunk
--- PASS: TestResourceFastlyFlattenSplunk (0.00s)
=== RUN   TestAccFastlyServiceV1_splunk_basic
--- PASS: TestAccFastlyServiceV1_splunk_basic (136.86s)
=== RUN   TestAccFastlyServiceV1_splunk_default
--- PASS: TestAccFastlyServiceV1_splunk_default (57.11s)
=== RUN   TestAccFastlyServiceV1_splunk_env
--- PASS: TestAccFastlyServiceV1_splunk_env (62.04s)
=== RUN   TestResourceFastlyFlattenSumologic
--- PASS: TestResourceFastlyFlattenSumologic (0.00s)
=== RUN   TestAccFastlyServiceV1_sumologic
--- PASS: TestAccFastlyServiceV1_sumologic (133.06s)
=== RUN   TestResourceFastlyFlattenSyslog
--- PASS: TestResourceFastlyFlattenSyslog (0.00s)
=== RUN   TestAccFastlyServiceV1_syslog_basic
--- PASS: TestAccFastlyServiceV1_syslog_basic (124.61s)
=== RUN   TestAccFastlyServiceV1_syslog_formatVersion
--- PASS: TestAccFastlyServiceV1_syslog_formatVersion (63.49s)
=== RUN   TestResourceFastlyFlattenDomains
--- PASS: TestResourceFastlyFlattenDomains (0.00s)
=== RUN   TestResourceFastlyFlattenBackend
--- PASS: TestResourceFastlyFlattenBackend (0.00s)
=== RUN   TestAccFastlyServiceV1_updateDomain
--- PASS: TestAccFastlyServiceV1_updateDomain (159.93s)
=== RUN   TestAccFastlyServiceV1_updateBackend
--- PASS: TestAccFastlyServiceV1_updateBackend (130.82s)
=== RUN   TestAccFastlyServiceV1_updateInvalidBackend
--- PASS: TestAccFastlyServiceV1_updateInvalidBackend (67.34s)
=== RUN   TestAccFastlyServiceV1_basic
--- PASS: TestAccFastlyServiceV1_basic (122.89s)
=== RUN   TestAccFastlyServiceV1_disappears
--- PASS: TestAccFastlyServiceV1_disappears (23.78s)
=== RUN   TestAccFastlyServiceV1_defaultTTL
--- PASS: TestAccFastlyServiceV1_defaultTTL (198.64s)
=== RUN   TestResourceFastlyFlattenVCLs
--- PASS: TestResourceFastlyFlattenVCLs (0.00s)
=== RUN   TestAccFastlyServiceV1_VCL_basic
--- PASS: TestAccFastlyServiceV1_VCL_basic (128.72s)
=== RUN   TestValidateLoggingFormatVersion
=== RUN   TestValidateLoggingFormatVersion/5
=== RUN   TestValidateLoggingFormatVersion/0
=== RUN   TestValidateLoggingFormatVersion/1
=== RUN   TestValidateLoggingFormatVersion/2
=== RUN   TestValidateLoggingFormatVersion/3
=== RUN   TestValidateLoggingFormatVersion/4
--- PASS: TestValidateLoggingFormatVersion (0.00s)
    --- PASS: TestValidateLoggingFormatVersion/5 (0.00s)
    --- PASS: TestValidateLoggingFormatVersion/0 (0.00s)
    --- PASS: TestValidateLoggingFormatVersion/1 (0.00s)
    --- PASS: TestValidateLoggingFormatVersion/2 (0.00s)
    --- PASS: TestValidateLoggingFormatVersion/3 (0.00s)
    --- PASS: TestValidateLoggingFormatVersion/4 (0.00s)
=== RUN   TestValidateLoggingMessageType
=== RUN   TestValidateLoggingMessageType/classic
=== RUN   TestValidateLoggingMessageType/loggly
=== RUN   TestValidateLoggingMessageType/logplex
=== RUN   TestValidateLoggingMessageType/blank
=== RUN   TestValidateLoggingMessageType/CLASSIC
=== RUN   TestValidateLoggingMessageType/LOGGLY
=== RUN   TestValidateLoggingMessageType/LOGPLEX
=== RUN   TestValidateLoggingMessageType/BLANK
--- PASS: TestValidateLoggingMessageType (0.00s)
    --- PASS: TestValidateLoggingMessageType/classic (0.00s)
    --- PASS: TestValidateLoggingMessageType/loggly (0.00s)
    --- PASS: TestValidateLoggingMessageType/logplex (0.00s)
    --- PASS: TestValidateLoggingMessageType/blank (0.00s)
    --- PASS: TestValidateLoggingMessageType/CLASSIC (0.00s)
    --- PASS: TestValidateLoggingMessageType/LOGGLY (0.00s)
    --- PASS: TestValidateLoggingMessageType/LOGPLEX (0.00s)
    --- PASS: TestValidateLoggingMessageType/BLANK (0.00s)
=== RUN   TestValidateLoggingPlacement
=== RUN   TestValidateLoggingPlacement/none
=== RUN   TestValidateLoggingPlacement/waf_debug
=== RUN   TestValidateLoggingPlacement/NONE
=== RUN   TestValidateLoggingPlacement/WAF_DEBUG
--- PASS: TestValidateLoggingPlacement (0.00s)
    --- PASS: TestValidateLoggingPlacement/none (0.00s)
    --- PASS: TestValidateLoggingPlacement/waf_debug (0.00s)
    --- PASS: TestValidateLoggingPlacement/NONE (0.00s)
    --- PASS: TestValidateLoggingPlacement/WAF_DEBUG (0.00s)
=== RUN   TestValidateDirectorQuorum
=== RUN   TestValidateDirectorQuorum/0
=== RUN   TestValidateDirectorQuorum/55
=== RUN   TestValidateDirectorQuorum/100
=== RUN   TestValidateDirectorQuorum/-1
=== RUN   TestValidateDirectorQuorum/101
=== RUN   TestValidateDirectorQuorum/150
--- PASS: TestValidateDirectorQuorum (0.00s)
    --- PASS: TestValidateDirectorQuorum/0 (0.00s)
    --- PASS: TestValidateDirectorQuorum/55 (0.00s)
    --- PASS: TestValidateDirectorQuorum/100 (0.00s)
    --- PASS: TestValidateDirectorQuorum/-1 (0.00s)
    --- PASS: TestValidateDirectorQuorum/101 (0.00s)
    --- PASS: TestValidateDirectorQuorum/150 (0.00s)
=== RUN   TestValidateDirectorType
=== RUN   TestValidateDirectorType/3
=== RUN   TestValidateDirectorType/4
=== RUN   TestValidateDirectorType/5
=== RUN   TestValidateDirectorType/0
=== RUN   TestValidateDirectorType/1
=== RUN   TestValidateDirectorType/2
--- PASS: TestValidateDirectorType (0.00s)
    --- PASS: TestValidateDirectorType/3 (0.00s)
    --- PASS: TestValidateDirectorType/4 (0.00s)
    --- PASS: TestValidateDirectorType/5 (0.00s)
    --- PASS: TestValidateDirectorType/0 (0.00s)
    --- PASS: TestValidateDirectorType/1 (0.00s)
    --- PASS: TestValidateDirectorType/2 (0.00s)
=== RUN   TestValidateConditionType
=== RUN   TestValidateConditionType/REQUEST
=== RUN   TestValidateConditionType/RESPONSE
=== RUN   TestValidateConditionType/CACHE
=== RUN   TestValidateConditionType/request
=== RUN   TestValidateConditionType/response
=== RUN   TestValidateConditionType/cache
--- PASS: TestValidateConditionType (0.00s)
    --- PASS: TestValidateConditionType/REQUEST (0.00s)
    --- PASS: TestValidateConditionType/RESPONSE (0.00s)
    --- PASS: TestValidateConditionType/CACHE (0.00s)
    --- PASS: TestValidateConditionType/request (0.00s)
    --- PASS: TestValidateConditionType/response (0.00s)
    --- PASS: TestValidateConditionType/cache (0.00s)
=== RUN   TestValidateHeaderAction
=== RUN   TestValidateHeaderAction/set
=== RUN   TestValidateHeaderAction/append
=== RUN   TestValidateHeaderAction/delete
=== RUN   TestValidateHeaderAction/regex
=== RUN   TestValidateHeaderAction/regex_repeat
=== RUN   TestValidateHeaderAction/SET
=== RUN   TestValidateHeaderAction/APPEND
=== RUN   TestValidateHeaderAction/DELETE
=== RUN   TestValidateHeaderAction/REGEX
=== RUN   TestValidateHeaderAction/REGEX_REPEAT
--- PASS: TestValidateHeaderAction (0.00s)
    --- PASS: TestValidateHeaderAction/set (0.00s)
    --- PASS: TestValidateHeaderAction/append (0.00s)
    --- PASS: TestValidateHeaderAction/delete (0.00s)
    --- PASS: TestValidateHeaderAction/regex (0.00s)
    --- PASS: TestValidateHeaderAction/regex_repeat (0.00s)
    --- PASS: TestValidateHeaderAction/SET (0.00s)
    --- PASS: TestValidateHeaderAction/APPEND (0.00s)
    --- PASS: TestValidateHeaderAction/DELETE (0.00s)
    --- PASS: TestValidateHeaderAction/REGEX (0.00s)
    --- PASS: TestValidateHeaderAction/REGEX_REPEAT (0.00s)
=== RUN   TestValidateHeaderType
=== RUN   TestValidateHeaderType/request
=== RUN   TestValidateHeaderType/fetch
=== RUN   TestValidateHeaderType/cache
=== RUN   TestValidateHeaderType/response
=== RUN   TestValidateHeaderType/REQUEST
=== RUN   TestValidateHeaderType/FETCH
=== RUN   TestValidateHeaderType/CACHE
=== RUN   TestValidateHeaderType/RESPONSE
--- PASS: TestValidateHeaderType (0.00s)
    --- PASS: TestValidateHeaderType/request (0.00s)
    --- PASS: TestValidateHeaderType/fetch (0.00s)
    --- PASS: TestValidateHeaderType/cache (0.00s)
    --- PASS: TestValidateHeaderType/response (0.00s)
    --- PASS: TestValidateHeaderType/REQUEST (0.00s)
    --- PASS: TestValidateHeaderType/FETCH (0.00s)
    --- PASS: TestValidateHeaderType/CACHE (0.00s)
    --- PASS: TestValidateHeaderType/RESPONSE (0.00s)
=== RUN   TestValidateSnippetType
=== RUN   TestValidateSnippetType/init
=== RUN   TestValidateSnippetType/recv
=== RUN   TestValidateSnippetType/hit
=== RUN   TestValidateSnippetType/miss
=== RUN   TestValidateSnippetType/pass
=== RUN   TestValidateSnippetType/fetch
=== RUN   TestValidateSnippetType/error
=== RUN   TestValidateSnippetType/deliver
=== RUN   TestValidateSnippetType/log
=== RUN   TestValidateSnippetType/none
=== RUN   TestValidateSnippetType/INIT
=== RUN   TestValidateSnippetType/RECV
=== RUN   TestValidateSnippetType/HIT
=== RUN   TestValidateSnippetType/MISS
=== RUN   TestValidateSnippetType/PASS
=== RUN   TestValidateSnippetType/FETCH
=== RUN   TestValidateSnippetType/ERROR
=== RUN   TestValidateSnippetType/DELIVER
=== RUN   TestValidateSnippetType/LOG
=== RUN   TestValidateSnippetType/NONE
--- PASS: TestValidateSnippetType (0.00s)
    --- PASS: TestValidateSnippetType/init (0.00s)
    --- PASS: TestValidateSnippetType/recv (0.00s)
    --- PASS: TestValidateSnippetType/hit (0.00s)
    --- PASS: TestValidateSnippetType/miss (0.00s)
    --- PASS: TestValidateSnippetType/pass (0.00s)
    --- PASS: TestValidateSnippetType/fetch (0.00s)
    --- PASS: TestValidateSnippetType/error (0.00s)
    --- PASS: TestValidateSnippetType/deliver (0.00s)
    --- PASS: TestValidateSnippetType/log (0.00s)
    --- PASS: TestValidateSnippetType/none (0.00s)
    --- PASS: TestValidateSnippetType/INIT (0.00s)
    --- PASS: TestValidateSnippetType/RECV (0.00s)
    --- PASS: TestValidateSnippetType/HIT (0.00s)
    --- PASS: TestValidateSnippetType/MISS (0.00s)
    --- PASS: TestValidateSnippetType/PASS (0.00s)
    --- PASS: TestValidateSnippetType/FETCH (0.00s)
    --- PASS: TestValidateSnippetType/ERROR (0.00s)
    --- PASS: TestValidateSnippetType/DELIVER (0.00s)
    --- PASS: TestValidateSnippetType/LOG (0.00s)
    --- PASS: TestValidateSnippetType/NONE (0.00s)
=== RUN   TestValidateDictionaryItemMaxSize
=== RUN   TestValidateDictionaryItemMaxSize/Ten_thousand_dictionary_items
=== RUN   TestValidateDictionaryItemMaxSize/Ten_thousand_and_one_dictionary_items
=== RUN   TestValidateDictionaryItemMaxSize/Ten_hundred_dictionary_items
--- PASS: TestValidateDictionaryItemMaxSize (0.01s)
    --- PASS: TestValidateDictionaryItemMaxSize/Ten_thousand_dictionary_items (0.00s)
    --- PASS: TestValidateDictionaryItemMaxSize/Ten_thousand_and_one_dictionary_items (0.00s)
    --- PASS: TestValidateDictionaryItemMaxSize/Ten_hundred_dictionary_items (0.00s)
PASS
ok      github.com/terraform-providers/terraform-provider-fastly/fastly 4595.587s
?       github.com/terraform-providers/terraform-provider-fastly/version        [no test files]

```

</p>
</details>

philippschulte commented on website/docs/r/service_dictionary_items_v1.html.markdown:
````diff
@@ -0,0 +1,203 @@
+}
+```
+
+###Supporting API and UI dictionary updates with ignore_changes
````
Please add a blank after the third `#`.

philippschulte commented on website/docs/r/service_dictionary_items_v1.html.markdown:
````diff
@@ -0,0 +1,203 @@
+Basic usage:
+
+```hcl
+
````
Please remove the extra line.

philippschulte commented on website/docs/r/service_dictionary_items_v1.html.markdown:
```diff
@@ -0,0 +1,203 @@
+        key2: "value2"
+    }
+}
+
```
Please remove the extra line.

philippschulte commented on website/docs/r/service_dictionary_items_v1.html.markdown:
````diff
@@ -0,0 +1,203 @@
+Complex object usage:
+
+```hcl
+
````
Please remove the extra line.

philippschulte commented on website/docs/r/service_dictionary_items_v1.html.markdown:
```diff
@@ -0,0 +1,203 @@
+  dictionary_id = {for d in fastly_service_v1.myservice.dictionary : d.name => d.dictionary_id}[var.mydict.name]
+  items = var.mydict.items
+}
+
```
Please remove the extra line.

philippschulte commented on website/docs/r/service_dictionary_items_v1.html.markdown:
```diff
@@ -0,0 +1,203 @@
+    for division in local.host_divisions:
+      division => format("%s.%s", division, local.host_base)
+  }
+
```
Please remove the extra line.

philippschulte commented on website/docs/r/service_dictionary_items_v1.html.markdown:
```diff
@@ -0,0 +1,203 @@
+    }
+}
+
+
```
Please remove the extra lines.

philippschulte reviewed (approved):
This looks like a solid and robust implementation for me! Well done!

Commit:
Adjusts the formatting around the examples and headings

trentrosenbaum commented:
Just to confirm that I have updated the doc with the format suggestions.

philippschulte reviewed (changes requested):
@trentrosenbaum I actually found one more thing while reviewing the ACL implementation!. The `dictionary` nested resource needs to be documented as well (fastly_service_v1). It's also important to document the exported ID [here](https://www.terraform.io/docs/providers/fastly/r/service_v1.html#attributes-reference). Thank you.

Commit:
Added defined below to dictionary nested block bullet point.

Commit:
Updates website documentation to detail dictionary_id as an export

trentrosenbaum commented:
@philippschulte I have updated the missing entry in the service_v1.html.markdown and I also updated the exported attributes as suggested at the bottom.

phamann commented on website/docs/r/service_dictionary_items_v1.html.markdown:
```diff
@@ -0,0 +1,196 @@
+
+### Supporting API and UI dictionary updates with ignore_changes
+
+The following example demonstrates how the lifecycle ignore_change field can be used to suppress updates against the 
```
Backticks around `ignore_changes` please for consistency

trentrosenbaum replied:
backticks have been added.

phamann commented on website/docs/r/service_dictionary_items_v1.html.markdown:
```diff
@@ -0,0 +1,196 @@
+Defines a map of Fastly dictionary items that can be used to populate a service dictionary.  This resource will populate a dictionary with the items and will track their state.
+
+
+~> **Warning:** Terraform will take precedence over any changes you make in the UI or API. Such changes are likely to be reversed if you run terraform again.  
```
Uppercase T on second Terraform

trentrosenbaum replied:
Uppercase T has been corrected.

phamann reviewed (approved):
👍 LGTM

Commit:
Corrects format details after review.

philippschulte reviewed (approved):
👍 LGTM
